Tests for ``100-matrix_mul``
===========================

Functions:
----------
1. ``matrix_mul(m_a, m_b)``
This function multiplies two matrices

`m_a` and `m_b` are both matrices(lists of lists)
To be multiplicable, the number of rows in `m_a` has to equal the number of columns in `m_b`

import the function from the module:
>>> mul = __import__('100-matrix_mul').matrix_mul

Tests
-----

Okay args:
>>> print(mul([[1, 2], [3, 4]], [[1, 2], [3, 4]]))
[[7, 10], [15, 22]]
>>> print(mul([[1.0, 2.0]], [[3, 4], [5, 6]]))
[[13.0, 16.0]]
>>> print(mul([[1, 2, 3]], [[1, 2], [2, 4], [3, 6]]))
[[14, 28]]

Missing args:
(No defaults, exception raised)
>>> print(mul())
Traceback (most recent call last):
    ...
TypeError: matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'
>>> print(mul([[1, 2, 3]]))
Traceback (most recent call last):
    ...
TypeError: matrix_mul() missing 1 required positional argument: 'm_b'


Bad arguments:
Not lists
>>> print(mul(1,[[2, 4]]))
Traceback (most recent call last):
    ...
TypeError: m_a must be a list
>>> print(mul([[2, 4]], (6, 7)))
Traceback (most recent call last):
    ...
TypeError: m_b must be a list

Not lists of lists
>>> print(mul([1, 2], [[1, 3], [4, 5]]))
Traceback (most recent call last):
    ...
TypeError: m_a must be a list of lists
>>> print(mul([[1, 3], [4, 5]], [5, 7]))
Traceback (most recent call last):
    ...
TypeError: m_b must be a list of lists

Empty lists
>>> print(mul([], [[12, 45]]))
Traceback (most recent call last):
    ...
ValueError: m_a can't be empty
>>> print(mul([[1, 2], [5, 7]], [[]]))
Traceback (most recent call last):
    ...
ValueError: m_b can't be empty

Non-numbers:
>>> print(mul([['2', 3, 4]],[[2, 3], [6,7], [8,9]]))
Traceback (most recent call last):
    ...
TypeError: m_a should contain only integers or floats
>>> print(mul([[2, 3, 4]],[['2', '5'], [6,7], [8,9]]))
Traceback (most recent call last):
    ...
TypeError: m_b should contain only integers or floats

Uneven row sizes:
>>> print(mul([[1, 2, 5], [6,7], [8,9]], [[1,2],[5,6]]))
Traceback (most recent call last):
    ...
TypeError: each row of m_a must be of the same size
>>> print(mul([[2, 3, 4]],[[1, 2, 5], [6,7], [8,9]]))
Traceback (most recent call last):
    ...
TypeError: each row of m_b must be of the same size

Cannot be multiplied:
>>> print(mul([[1, 2], [3, 4]], [[1, 2, 3], [10,20,30], [4,5,6]]))
Traceback (most recent call last):
    ...
ValueError: m_a and m_b can't be multiplied

Multiple violations:
Order of evaluation
list->list of lists->empty->numbers->size->multiplicable
>>> print(mul([[1, '2']], 6))
Traceback (most recent call last):
    ...
TypeError: m_b must be a list
